src/gafaelfawr/storage/token.py

```diff
@@ -0,0 +1,383 @@
+"""Storage for tokens."""
+
+from __future__ import annotations
+
+from datetime import datetime, timezone
+from typing import TYPE_CHECKING
+
+from gafaelfawr.exceptions import DeserializeException, DuplicateTokenNameError
+from gafaelfawr.models.token import TokenInfo, TokenType
+from gafaelfawr.schema.subtoken import Subtoken
+from gafaelfawr.schema.token import Token as SQLToken
+
+if TYPE_CHECKING:
+    from typing import List, Optional
+
+    from sqlalchemy.orm import Session
+    from structlog import BoundLogger
+
+    from gafaelfawr.models.token import Token, TokenData
+    from gafaelfawr.storage.base import RedisStorage
+
+__all__ = ["TokenDatabaseStore", "TokenRedisStore"]
+
+
+class TokenDatabaseStore:
+    """Stores and manipulates tokens in the database.
+
+    Tokens exist in both Redis and in the database.  Redis is the source of
+    truth for the validity of the token and the only data store that holds the
+    supplemental user information that will eventually be replaced by an
+    identity management system.  The database is the canonical store for
+    user-given token names and for the relationship between tokens.
+
+    Parameters
+    ----------
+    session : `sqlalchemy.orm.Session`
+        The underlying database session.
+    """
+
+    def __init__(self, session: Session) -> None:
+        self._session = session
+
+    def add(
+        self,
+        data: TokenData,
+        *,
+        token_name: Optional[str] = None,
+        service: Optional[str] = None,
+        parent: Optional[str] = None,
+    ) -> None:
+        """Store a new token.
+
+        Parameters
+        ----------
+        data : `gafaelfawr.models.token.TokenData`
+            The corresponding data.
+        token_name : `str`, optional
+            The human-given name for the token.
+        service : `str`, optional
+            The service for an internal token.
+        parent : `str`, optional
+            The key of the parent of this token.
+
+        Raises
+        ------
+        gafaelfawr.exceptions.DuplicateTokenNameError
+            The user already has a token by that name.
+        """
+        if token_name:
+            name_conflict = (
+                self._session.query(SQLToken.token)
+                .filter_by(username=data.username, token_name=token_name)
+                .scalar()
+            )
+            if name_conflict:
+                msg = f"Token name {token_name} already used"
+                raise DuplicateTokenNameError(msg)
+        new = SQLToken(
```

Comment:
If there's a race, you'll miss out on raising `DuplicateTokenNameError` for the unique key violation, right?

Reply:
This happens inside a transaction, so hopefully with most databases that will make the race at least unlikely, but yes, it's possible that some races will result in an uncaught SQL exception. If you know of a good technique to avoid those in general, I'd love to hear more; I've never found a great way of doing that and tend to live with a few narrow race conditions producing 500 errors in database-driven web applications if two servers race the creation of an object in just the right way.

Reply:
In this case, I don't think a race condition would cause database corruption or even improper behavior; it's only that the error returned could be less descriptive/friendly than it would be otherwise.  So no big deal.